Curvature/URI/URI.swift

```diff
@@ -22,26 +22,37 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-public struct URI {
-    public struct UserInfo {
-        public let username: String
-        public let password: String
+public struct URICredentials {
+    public let username: String
+    public let password: String
+    
+    public init(username: String, password: String) {
+        self.username = username
+        self.password = password
+    }
+}
 
-        public init(username: String, password: String) {
-            self.username = username
-            self.password = password
-        }
+extension URICredentials : Hashable {
+    public var hashValue: Int {
+        return username.hashValue ^ password.hashValue
     }
```

Comment:
We can keep this inside URI as this object only exists in the context of an URI. The name should also stay UserInfo, which is the official term in [RFC 3986](https://tools.ietf.org/html/rfc3986). You can refer to the inner struct outside the parent struct like this:

```swift
public struct URI {
    public struct UserInfo: Hashable {
        public let username: String
        public let password: String
        
        public init(username: String, password: String) {
            self.username = username
            self.password = password
        }
        
        public var hashValue: Int {
            return username.hashValue ^ password.hashValue
        }
    }
    // ...
}

public func ==(lhs: URI.UserInfo, rhs: URI.UserInfo) -> Bool {
    return lhs.hashValue == rhs.hashValue
}
```

Is there a specific reason you think `UserInfo` should be outside `URI`? My concern is that `UserInfo` is a rather used name in other contexts. Although if something happens, the user can namespace it to the framework, calling it by its long form: `Curvature.UserInfo`. But if we namespace it beforehand to `URI` it makes much more sense to call `URI.UserInfo` than `Curvature.UserInfo`.